PBS tests not failing when they should

milljm opened the pull request:
Set the exit code to what ever it was when the
application exited, as PBS saw it.

If PBS kills the job due to exceeding walltime,
a non-zero value is recorded, and the test will
now be labeled as failed (crash).

Closes #10476

Commit:
PBS tests not failing when they should

Set the exit code to what ever it was when the
application exited, as PBS saw it.

If PBS kills the job due to exceeding walltime,
a non-zero value is recorded, and the test will
now be labeled as failed (crash).

Closes #10476

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/144023/) on 8c927ad wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10477/site)

This comment will be updated on new commits.

milljm commented:
There is an issue with a string/integer comparison with this PR. Do not merge. I want to create a unittest case for walltime exceeded failures.

Commit:
Update comments, add unittest

Closes #10476

Commit:
Remove the import re module

Addressing pyflakes failures.

Closes #10476

milljm commented:
There is one more thing I wanted to add to this PR that I just remembered...

An exception can trigger when processResults() and the long running timer occur at the same time.
The timer (at just the right time), sets a pending status, to a job which has _just_ finished, but has yet to enter the status printing queue. This causes the scheduler to complain about an unfinished job:
```python
# Derived run needs to set a non-pending status of some sort.
if tester.isPending():
    raise SchedulerError('Derived Scheduler %s can not return a pending status!' % (self.__class__))
```

The simple fix, is to only allow the long running timer to set a status, if the current status is not a finished one.

milljm commented:
I wanted to add this last commit to _this_ PR, because the issue is most prevalent with PBS submitted jobs.

Commit:
Prevent long-running timer from interfering

There exists a chance that the long running timer might
set a pending status on a job which otherwise is in the
process of exiting (and has already set a non-pending
status). This change only allows the long running timer
to adjust statuses for tests that have not yet finished.

Closes #10476
